Sprint 14

Aleksandr0406 opened the pull request:
1. Создан новый проект Tracker, iOS 13.4 - минимально поддерживаемая версия;

2. Вёрстка экранов соответствует макету
 
3. Свёрстан Launch Screen приложения:

4. На главном экране — две вкладки: «Трекеры» и «Статистика»

5. Свёрстан главный экран «Трекеры»:

6. В проект добавлены структуры Tracker, TrackerCategory и TrackerRecord 

7. Реализованы экраны создания трекера

8. Реализован экран задания расписания для трекера

9. Приложение поддерживает устройства iPhone с iOS 13.4 или выше

Aleksandr0406 committed:
Добавлен лаунч скрин и главный экран

Aleksandr0406 committed:
Добавлены модели Tracker, TrackerCategory, TrackerRecord

Aleksandr0406 committed:
Добавлены следующие экраны: создание привычки и нерегулярного события и настроен главный экран на отображение созданных трекеров

Aleksandr0406 committed:
Реализована работа и анимация кнопки на трекере

NaR0RoW commented on Tracker.xcodeproj/xcuserdata/a1111.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:
Папку `xcuserdata` стоит игнорировать, кладя в `.gitignore`. В ней хранится информация о состоянии проекта (какой файл открыт, какой таргет выбран и т.п.), брейкпоинты и схемы для текущего пользователя `Xcode`.
Про `gitignore` можно прочитать вот [тут](https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files)

NaR0RoW commented on Tracker/Assets&LaunchSc&Info/Assets.xcassets/Images/Add_button.imageset/Add tracker@1x.png:
Картинки 1x лучше не добавлять. Они нужны были для iPhone без retina экранов, а последний такой - iPhone 5. Он не поддерживает минимальную версию учебного приложения, поэтому можно смело удалять, чтобы не занимать лишнюю память

NaR0RoW commented on Tracker/AppDelegate.swift:
```diff
@@ -9,9 +9,8 @@ import UIKit
 
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
```
Стоит пометить как `final`. Если класс не наследуется, то лучше пометить его как `final` для ускорения работы компилятора

NaR0RoW commented on Tracker/AppDelegate.swift:
```diff
@@ -20,6 +19,10 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
+//        let window = UIWindow()
+//        window.rootViewController = TabBarController()
+//        self.window = window
+//        self.window?.makeKeyAndVisible()
```
Лучше удалять закомментированный код перед review

NaR0RoW commented on Tracker/Models/Tracker.swift:
```diff
@@ -0,0 +1,17 @@
+//  Created by 1111 on 04.02.2025.
+//
+
+import Foundation
+import UIKit
```
`Foundation` можно не импортировать, т.к. он уже есть в `UIKit`

Aleksandr0406 replied:
убрал import Foundation везде где уже импортировал UIKit

NaR0RoW commented on Tracker/NewCategoryViewController/NewCategoryNameViewController.swift:
```diff
@@ -0,0 +1,84 @@
+//
+
+import Foundation
+import UIKit
```
Стоит везде убрать `import Foundation` если в файле уже есть импорт UIKit

NaR0RoW commented on Tracker/NewCategoryViewController/NewCategoryNameViewController.swift:
```diff
@@ -0,0 +1,84 @@
+import UIKit
+
+final class NewCategoryNameViewController: UIViewController, UITextFieldDelegate {
+    var clousure: ((String) -> ())!
```
Давай не будем использовать `force` операции

Aleksandr0406 replied:
изменил на var clousure: ((String) -> ())? и опционал по месту распаковал(в @objc private func didTapDoneButton())

NaR0RoW commented on Tracker/NewCategoryViewController/NewCategoryNameViewController.swift:
```diff
@@ -0,0 +1,84 @@
+        titleCategoryTextField.translatesAutoresizingMaskIntoConstraints = false
+        view.addSubview(titleCategoryTextField)
+        
+        titleCategoryTextField.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 24).isActive = true
+        titleCategoryTextField.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16).isActive = true
+        titleCategoryTextField.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16).isActive = true
+        titleCategoryTextField.heightAnchor.constraint(equalToConstant: 75).isActive = true
```
Можно такую форму записи использовать:
``` swift
NSLayoutConstraint.activate([
   titleCategoryTextField.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor, constant: 24),
   titleCategoryTextField.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16),
   titleCategoryTextField.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16),
   titleCategoryTextField.heightAnchor.constraint(equalToConstant: 75)
])
```

Aleksandr0406 replied:
да знаю, что есть такая форма записи- мне как то привычнее использовать активацию каждого констрейнта) Есть какая принципиальная разница в записях или только в меньшем кол-ве кода?

NaR0RoW replied:
Использование `NSLayoutConstraint.activate` считается предпочтительней:
1. `NSLayoutConstraint.activate` активирует все констрейнты одновременно, соот. это уменьшает количество обновлений layout движка 
2. `isActive = true` активирует констрейнты по одному, каждый раз вызывая лишнее обновление layout'a, что не круто, когда констрейнтов много
3. В целом запись `NSLayoutConstraint.activate` компактнее и ее легче читать, а `isActive = true` занимает больше строк

`isActive = true` лучше использовать, когда констрейнты устанавливаются динамически или когда нужно активировать / деактивировать констрейнты по одному в разных частях кода

Aleksandr0406 replied:
Изменил во всем проекте на NSLayoutConstraint.activate

NaR0RoW commented on Tracker/NewCategoryViewController/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,157 @@
+import UIKit
+
+final class NewCategoryViewController: UIViewController {
+    var clousure: ((String) -> ())!
```
Давай нигде не будем использовать `force` операции. В последующих спринтах использование `force unwrap` будет считаться критической ошибкой

Aleksandr0406 replied:
изменил на var clousure: ((String) -> ())? и опционал по месту распаковал( в func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath))

NaR0RoW commented on Tracker/NewCategoryViewController/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,157 @@
+    
+    @objc private func didTapAddCategoryButton() {
+        let viewcontroller = NewCategoryNameViewController()
+        viewcontroller.clousure = { nameCategory in
```
Для большей определенности стоит переименовать `closure` например на `onDoneButtonTapped`

Aleksandr0406 replied:
Все clousure в проекте переименовал, превратил в опционалы и распаковал по месту. Новые названия следующие:
1. NewCategoryNameViewController: clousure -> onDoneButtonTapped
2. NewCategoryViewController: clousure -> onAddCategoryButtonTapped
3. NewScheduleViewController: clousure -> onDoneButtonTapped
4. NewHabitViewController: clousure -> onAddHabbitButtonTapped
5. NewTrackerViewController: clousure -> onAddHabbitButtonTapped

NaR0RoW commented on Tracker/NewCategoryViewController/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,157 @@
+
+extension NewCategoryViewController: UITableViewDelegate, UITableViewDataSource {
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return names.count
```
В однострочных выражениях `return` можно опускать

Aleksandr0406 replied:
Пробежался по проекту и убрал return во всех допустимых местах

NaR0RoW commented on Tracker/NewCategoryViewController/NewCategoryViewController.swift:
```diff
@@ -0,0 +1,157 @@
+        tableView.cellForRow(at: indexPath)?.accessoryType = .checkmark
+        index = indexPath.row
+        guard let subtitleNameCategory = tableView.cellForRow(at: indexPath)?.textLabel?.text else { return }
+        clousure(subtitleNameCategory)
```
Ошибка в правописании:
`closure` -> `clousure`

Лучше включить [настройку](https://sarunw.com/posts/spell-checking-in-xcode/), чтобы `Xcode` подсвечивал ошибки в правописании

Aleksandr0406 replied:
Настройку включил, кложуры все переименовал

NaR0RoW commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -0,0 +1,429 @@
+    private var warningLabel: UILabel = UILabel()
+    private var titleHabitTextField: UITextField = UITextField()
+    private var cancelButton: UIButton = UIButton()
+    private var addHabbitButton: UIButton = UIButton()
```
Ошибка в правописании:
`addHabbitButton` -> `addHabitButton`

NaR0RoW commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -0,0 +1,429 @@
+        
+        guard let numbersOfLetters = titleHabitTextField.text?.count else { return }
+        
+        if numbersOfLetters == 38 {
```
Лучше избегать магических чисел (число 38) и использовать константы. Имя константы будет раскрывать предназначение числа

Aleksandr0406 replied:
Переименовал в enableNumberOfTypingLettersInTextField, в проекте изменил

NaR0RoW commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -0,0 +1,429 @@
+        
+        guard let numbersOfLetters = titleHabitTextField.text?.count else { return }
+        
+        if numbersOfLetters == 38 {
+            warningLabel.isHidden = false
+        } else {
+            warningLabel.isHidden = true
+        }
```
Можно в одну строчку выражение записать:
``` swift
warningLabel.isHidden = numbersOfLetters == 38
```

Aleksandr0406 replied:
изменил на
warningLabel.isHidden = !(numbersOfTypingLetters == enableNumberOfTypingLettersInTextField)

NaR0RoW commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -0,0 +1,429 @@
+        categoryAndScheduleTable.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 16).isActive = true
+        categoryAndScheduleTable.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -16).isActive = true
+        
+        if categoriesAndSchedule.count > 1 {
+            categoryAndScheduleTable.heightAnchor.constraint(equalToConstant: 150).isActive = true
+        } else {
+            categoryAndScheduleTable.heightAnchor.constraint(equalToConstant: 75).isActive = true
+        }
```
Можно тернарый оператор использовать:
``` swift
let padding: CGFloat = categoriesAndSchedule.count > 1 ? 150 : 75
categoryAndScheduleTable.heightAnchor.constraint(equalToConstant: padding).isActive = true
```

Aleksandr0406 replied:
а padding это норм название или нужно свое какое то придумывать в таких кейсах?

NaR0RoW replied:
Я в основном на `SwiftUI` пишу и там есть модификатор для настройки отступов (так и называется `padding`), поэтому и предложил такое название. Но конечно можно и свое название придумать, например, `constraintValue`. Наверное оно даже лучше будет, т.к. мы с `UIKit` работаем, а не `SwiftUI`

NaR0RoW commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -0,0 +1,429 @@
+        return numberOfSections
+    }
+    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        return (section == 0) ? emojies.count : colors.count
```
Лучше не использовать круглые скобки, если это явно не требуется

NaR0RoW commented on Tracker/NewHabbitViewController/NewHabitViewController.swift:
```diff
@@ -0,0 +1,429 @@
+            id = ""
+        }
+        
+        let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: id, for: indexPath) as! NewHabitCollectionSupplementaryView
```
Давай не будем force cast'ить

Aleksandr0406 replied:
изменил на 

        guard let headerView = collectionView.dequeueReusableSupplementaryView(ofKind: kind, withReuseIdentifier: id, for: indexPath) as? NewHabitCollectionSupplementaryView else { return UICollectionReusableView() }

NaR0RoW commented on Tracker/NewScheduleViewController/NewScheduleViewController.swift:
```diff
@@ -0,0 +1,125 @@
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        let cell: UITableViewCell
+        
+        if let newSheduleCell = tableView.dequeueReusableCell(withIdentifier: NewScheduleTableViewCell.cellIdentifier, for: indexPath) as? NewScheduleTableViewCell {
```
`newSheduleCell` -> `newScheduleCell`

NaR0RoW commented on Tracker/TrackersViewController/NewTrackerViewController.swift:
```diff
@@ -0,0 +1,90 @@
+import UIKit
+
+final class NewTrackerViewController: UIViewController {
+    var clousure: (((String, String, [String]) -> ()))!
```
Лучше дать более осмысленное название

Aleksandr0406 replied:
Переименовал все кложуры в проекте( выше более полный комментарий оставил в похожем замечании)

NaR0RoW commented on Tracker/TrackersViewController/TrackersCollectionCell.swift:
```diff
@@ -0,0 +1,166 @@
+    }
+    
+    private func plusOneDay(_ oneDay: Int) -> String {
+        var sumDay: Int = 0
```
`sumDay` можно не создавать

Aleksandr0406 replied:
ну да) изменил всю функцию на

    private func plusOneDay(_ oneDay: Int) -> String {
        "\(oneDay) день"
    }

саму функцию оставляю для большей определенности или лучше убрать?

NaR0RoW replied:
Думаю лучше убрать и поменять название константы для большей определенности, например:
``` swift
let completedDaysString: String = "\(completedDays) день"
```

NaR0RoW commented on Tracker/TrackersViewController/TrackersViewController.swift:
```diff
@@ -0,0 +1,364 @@
+                backgroundImage.isHidden = true
+                backgroundTextLabel.isHidden = true
+            }
+        }
```
Можно проще:
``` swift
let isEmpty = visibleCategories.allSatisfy { $0.trackers.isEmpty }
trackersCollectionView.isHidden = isEmpty
backgroundImage.isHidden = !isEmpty
backgroundTextLabel.isHidden = !isEmpty
```

NaR0RoW commented on Tracker/TrackersViewController/TrackersViewController.swift:
```diff
@@ -0,0 +1,364 @@
+    
+    //MARK: Helpers
+    private func convertSavedDaysToNumbersOfWeekend(_ savedDays: [String]) {
+        for day in savedDays {
+            if day == "Понедельник" {
+                savedDayNumberOfWeekend.append(2)
+            }
+            
+            if day == "Вторник" {
+                savedDayNumberOfWeekend.append(3)
+            }
+            
+            if day == "Среда" {
+                savedDayNumberOfWeekend.append(4)
+            }
+            
+            if day == "Четверг" {
+                savedDayNumberOfWeekend.append(5)
+            }
+            
+            if day == "Пятница" {
+                savedDayNumberOfWeekend.append(6)
+            }
+            
+            if day == "Суббота" {
+                savedDayNumberOfWeekend.append(7)
+            }
+            
+            if day == "Воскресенье" {
+                savedDayNumberOfWeekend.append(1)
+            }
```
Можно таким образом записать:
``` swift
let dayNumbers: [String: Int] = [
    "Понедельник": 2, "Вторник": 3, "Среда": 4,
    "Четверг": 5, "Пятница": 6, "Суббота": 7, "Воскресенье": 1
]
savedDayNumberOfWeekend = savedDays.compactMap { dayNumbers[$0] }
```

Aleksandr0406 replied:
спасибо, не додумался)

Aleksandr0406 committed:
Добавил в .gitignore папку xcuserdata, изменил мутабельность свойства trackers (TrackerCategory) на let, удалены лишние return

Aleksandr0406 commented:
@NaR0RoW привет, спасибо за ревью! 

1. По поводу краша: не понимаю как его устранить- поискал в инете, нашел на стэк оверфлоу похожий вопрос(https://stackoverflow.com/questions/78917217/error-dequeuing-custom-uicollectionviewcell), и что эта проблема появилась с 16 Xcode и вроде как связана с вызовом метода dequeueReusableSupplementaryView или dequeueReusableCell вне методов настройки ячейки или футеров/хэдеров. У меня эти методы(dequeue......) как раз используются только внутри настройки ячейки или футеров/хэдеров. Исходя из скрина ошибки, в моем проекте это связано именно с хэдером(относится к коллекции с эмоджи и цветами в NewHabitViewController), опять же метод dequeueReusableSupplementaryView  вызывается один раз и внутри func collectionView(_ collectionView: UICollectionView, viewForSupplementaryElementOfKind kind: String, at indexPath: IndexPath) -> UICollectionReusableView). У меня Xcode 15.4 и этой ошибки нет. 16 версию не ставлю, так как глючит. Можешь помочь?
 
2. По поводу иммутабельности свойства trackers структуры TrackerCategory. Поправил на let и изменил сохранение новой привычки, и получилось как то некрасиво - можешь прокомментировать пжл(решение в методе func updateTrackers(_ savedHabitName: String, _ savedCategoryName: String, _ savedDays: [String]), класс TrackersViewController

3. По поводу типа данных id- пока не разбирался с типом UUID, оставил стринг. В задании написано рекомендуем. На сколько критично сейчас это исправлять?

Дозалил обновления, буду ждать обратной связи.

NaR0RoW commented on .gitignore:
```diff
@@ -0,0 +1 @@
+/Users/a1111/Desktop/dev/projects/Tracker/Tracker/Tracker.xcodeproj/xcuserdata
```
Лучше таким образом записать:
`*.xcuserdata/`
Мы таким образом игнорируем все файлы `xcsuerdata` для всех пользователей и не привязываемся к конкретному пути

NaR0RoW commented:
1. Тут лучше к наставнику обратиться, полагаю, что в последнее время они часто встречаются с этим вопросом. 

2. Да, в целом нормально, но обычно внутри структуры добавляют метод для изменения самой структуры:
``` swift
struct TrackerCategory {
    let name: String
    let trackers: [Tracker]

    func updated(with newTrackers: [Tracker]) -> TrackerCategory {
        TrackerCategory(name: name, trackers: newTrackers)
    }
}
```
Ипользование:
``` swift
if savedCategoryName == categories[categoryIndex].name {
   let newTracker = Tracker(
      name: savedHabitName,
      color: .green,
      emoji: "🙂",
      schedule: savedDayNumberOfWeekend
   )
   categories[categoryIndex] = categories[categoryIndex].updated(with: categories[categoryIndex].trackers + [newTracker])
   break
}
```

3. В целом это не критичный момент, но рекомендую [UUID](https://developer.apple.com/documentation/foundation/uuid) использовать

`UUID (Universally Unique Identifier)` - это уникальный индификатор, обычно он как-то так выглядит:
`550e8400-e29b-41d4-a716-446655440000`

В Swift его можно таким образом задать, если нужна строка:

``` swift
let uuid = UUID().uuidString
print(uuid) // Пример: "A8B3C1D2-E4F5-6789-0ABC-DEF123456789"
```

Можно прямо внутри `Tracker` поменять:
``` swift
struct Tracker {
    let id: UUID = UUID()
    ...
}
```

Aleksandr0406 committed:
Изменена логика расчета высоты хэдеров(эмоджи и цвета) у коллекции на экране создании привычки; изменена реализация констрейнтов

Aleksandr0406 committed:
Изменил тип свойства id структуры Tracker на UUID; удалил функцию plusOneDay (не нужна)

Aleksandr0406 committed:
Удалил ненужные пробелы и отступы

Aleksandr0406 committed:
no message

Aleksandr0406 committed:
Исправлен расчет высоты хэдера коллеции на главном экране

Aleksandr0406 committed:
no message

Aleksandr0406 commented:
@NaR0RoW привет,

1. Вроде исправил: проблема была в реализации метода     func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize, который отвечает за расчет высоты хэдера коллекций. Хэдеры у меня сейчас настроены для двух экранов: главном и где создается привычка, оба поправил. Посмотри пжл, запускается ?

2. Принял к сведению, оставил пока мою реализацию

3. Изменил тип id на UUID и все поправил в проекте

 Перезалил, буду ждать ос

NaR0RoW commented:
@Aleksandr0406 Привет!
Приложение больше не падет. Перезапроси пожалуйста ревью на платформе

NaR0RoW commented:
Трекер нерегулярного события не добавляется на главный экран:

https://github.com/user-attachments/assets/55100708-581a-4dec-ae77-632f465d0964

Aleksandr0406 committed:
feat: добавить отступы в TextField'ы

В TextField на экране создания привычки/нерегулярного события и на экране создания названия категории добавлены отступы

Aleksandr0406 committed:
fix: изменить шрифт заглушки с жирного на обычный на экране создания новой категории

Aleksandr0406 committed:
refactor: удалить лишние принты и комменты

Aleksandr0406 committed:
refactor: общий рефакт в проекте

В расширениях к контроллерам, касающихся коллекций и таблиц, разделил методы на соответствующие протоколы

Aleksandr0406 commented:
@NaR0RoW привет, возвращаю с доработки:

Все три крита поправил- посмотри подходят ли решения?

По поводу рекомендации: лучше будет добавить на экран UIScrollView и в него уже все элементы положить или как то по другому? Пока не реализовывал

По поводу размышлений:

1. Про Conventional Commits прочитал, стал применять. Спасибо
2. Swift Style Guide читаю, в процессе, тоже спасибо. Пользуюсь параллельно вот этим гайдом(https://github.com/RedMadRobot/RMR-swift-style-guide#code-organization)

NaR0RoW commented:
@Aleksandr0406 
Посмотрел код - все хорошо. 

Да, стоит добавить `UIScrollView` как базовый контейнер, внутри добавить `UIStackView` (или `UIView`), который будет содержать все элементы (кнопки, коллекции и т. д.)

Круто, что начал использовать Conventional Commits 🔥
